### PR TITLE
Restart(Source|Flow|Sink): Configurable stream restart deadline

### PR DESCRIPTION
**Pull request by ismaelhamed**
Port [#29591](https://github.com/akka/akka/pull/29591)

This PR also adds a few missing tests, and attempts to fix the racy ones.

### CODE REVIEW

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs [hunk 1]
```diff
@@ -113,7 +125,8 @@ public void A_restart_with_backoff_source_should_restart_on_failure()
             }, Materializer);
         }
 
-        [Fact(Skip ="Racy")]
+        //[Fact(Skip ="Racy")]
+        [Fact]
```
Comment (ismaelhamed) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:129]: Racy, seems fixed now.

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs [hunk 2]
```diff
@@ -140,7 +155,8 @@ public void A_restart_with_backoff_source_should_backoff_before_restart()
...
```
Comment (ismaelhamed) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:159]: Racy, seems fixed now.

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs [hunk 3]
```diff
@@ -411,30 +458,30 @@ public void A_restart_with_backoff_sink_should_restart_on_cancellation()
             }, Materializer);
         }
 
-        [Fact(Skip ="Racy")]
+        //[Fact(Skip = "Racy")]
+        [Fact]
```
Comment (ismaelhamed) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:462]: Racy, seems fixed now.

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs [hunk 4]
```diff
@@ -444,46 +491,46 @@ public void A_restart_with_backoff_sink_should_backoff_before_restart()
...
```
Comment (ismaelhamed) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:495]: Racy, seems fixed now.

File: src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs [hunk 5]
```diff
@@ -708,31 +846,31 @@ public void A_restart_with_backoff_flow_should_restart_on_failure()
             created.Current.Should().Be(2);
         }
 
-        [Fact(Skip ="Racy")]
+        //[Fact(Skip = "Racy")]
+        [Fact]
```
Comment (ismaelhamed) [inline src/core/Akka.Streams.Tests/Dsl/RestartSpec.cs:850]: Racy, seems fixed now.

**Comment by ismaelhamed** [comment]
If racy tests are really fixed now, I just need to clean this PR up before it can be merged.

**Comment by ismaelhamed** [comment]
@Aaronontheweb this PR is ready now

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**